// static-jade-brunch/lib/index.js
// Generated by CoffeeScript 1.3.3
(function() {
  var StaticJadeCompiler, color, extend, fileWriter, fromJade2Html, fs, growl, isArray, jade, loadPackages, logError, mkdirp, sysPath,
    __indexOf = [].indexOf || function(item) { for (var i = 0, l = this.length; i < l; i++) { if (i in this && this[i] === item) return i; } return -1; };

  jade = require('jade');

  sysPath = require('path');

  mkdirp = require('mkdirp');

  fs = require('fs');

  color = require("ansi-color").set;

  growl = require('growl');

  fromJade2Html = function(jadeFilePath, config, callback) {
    try {
      return fs.readFile(jadeFilePath, function(err, data) {
        var content, foo, res, _ref, _ref1;
        content = jade.compile(data, {
          compileDebug: false,
          filename: jadeFilePath,
          pretty: !!((_ref = config.plugins) != null ? (_ref1 = _ref.jade) != null ? _ref1.pretty : void 0 : void 0)
        });
        foo = function() {};
        res = content(foo);
        return callback(err, res);
      });
    } catch (err) {
      return callback(err);
    }
  };

  logError = function(err, title) {
    if (!(title != null)) {
      title = 'Brunch jade error';
    }
    if (err != null) {
      console.log(color(err, "red"));
      return growl(err, {
        title: title
      });
    }
  };

  fileWriter = function(newFilePath) {
    return function(err, content) {
      var dirname;
      if (err != null) {
        throw err;
      }
      if (!(content != null)) {
        return;
      }
      dirname = sysPath.dirname(newFilePath);
      return mkdirp(dirname, '0775', function(err) {
        if (err != null) {
          throw err;
        }
        return fs.writeFile(newFilePath, content, function(err) {
          if (err != null) {
            throw err;
          }
        });
      });
    };
  };

  isArray = function(obj) {
    return !!(obj && obj.concat && obj.unshift && !obj.callee);
  };

  extend = function(object, properties) {
    Object.keys(properties).forEach(function(key) {
      return object[key] = properties[key];
    });
    return object;
  };

  loadPackages = function(rootPath, callback) {
    var nodeModules;
    rootPath = sysPath.resolve(rootPath);
    nodeModules = "" + rootPath + "/node_modules";
    return fs.readFile(sysPath.join(rootPath, 'package.json'), function(error, data) {
      var deps, json, plugins, _ref;
      if (error != null) {
        return callback(error);
      }
      json = JSON.parse(data);
      deps = Object.keys(extend((_ref = json.devDependencies) != null ? _ref : {}, json.dependencies));
      try {
        plugins = deps.map(function(dependency) {
          return require("" + nodeModules + "/" + dependency);
        });
      } catch (err) {
        error = err;
      }
      return callback(error, plugins);
    });
  };

  module.exports = StaticJadeCompiler = (function() {

    StaticJadeCompiler.prototype.brunchPlugin = true;

    StaticJadeCompiler.prototype.type = 'template';

    StaticJadeCompiler.prototype.extension = ".jade";

    function StaticJadeCompiler(config) {
      var _ref, _ref1, _ref2, _ref3, _ref4, _ref5;
      this.config = config;
      this.extension = (_ref = (_ref1 = this.config.plugins) != null ? (_ref2 = _ref1.static_jade) != null ? _ref2.extension : void 0 : void 0) != null ? _ref : ".jade";
      this.relAssetPath = (_ref3 = (_ref4 = this.config.plugins) != null ? (_ref5 = _ref4.static_jade) != null ? _ref5.asset : void 0 : void 0) != null ? _ref3 : "app/assets";
      mkdirp.sync(this.relAssetPath);
      StaticJadeCompiler.prototype.extension = this.extension;
      StaticJadeCompiler.prototype.config = this.config;
      loadPackages(process.cwd(), function(error, packages) {
        var errmsg, p;
        if (error != null) {
          throw error;
        }
        if (__indexOf.call((function() {
          var _i, _len, _results;
          _results = [];
          for (_i = 0, _len = packages.length; _i < _len; _i++) {
            p = packages[_i];
            _results.push(p.name);
          }
          return _results;
        })(), "JadeCompiler") < 0) {
          error = "`jade-brunch` plugin needed by `static-jade-brunch` \ndoesn't seems to be present.";
          logError(error, 'Brunch plugin error');
          errmsg = "* Check that package.json contain the `jade-brunch` plugin\n* Check that it is correctly installed by using `npm list`";
          console.log(color(errmsg, "red"));
          throw error;
        }
      });
    }

    StaticJadeCompiler.prototype.isFileToCompile = function(filePath) {
      var fileDir, fileName, p, positivePaths, _ref, _ref1;
      if ((((_ref = this.config.plugins) != null ? (_ref1 = _ref.static_jade) != null ? _ref1.path : void 0 : void 0) != null)) {
        if (isArray(this.config.plugins.static_jade.path)) {
          fileDir = sysPath.dirname(filePath);
          positivePaths = (function() {
            var _i, _len, _ref2, _results;
            _ref2 = this.config.plugins.static_jade.path;
            _results = [];
            for (_i = 0, _len = _ref2.length; _i < _len; _i++) {
              p = _ref2[_i];
              if (p.test(fileDir)) {
                _results.push(p);
              }
            }
            return _results;
          }).call(this);
          if (positivePaths.length === 0) {
            return false;
          }
        }
      }
      fileName = sysPath.basename(filePath);
      return fileName.slice(-this.extension.length) === this.extension;
    };

    StaticJadeCompiler.prototype.getHtmlFilePath = function(jadeFilePath, relAssetPath) {
      var newpath, relativeFilePath, relativeFilePathParts, util;
      util = require('util');
      relativeFilePathParts = jadeFilePath.split(sysPath.sep);
      relativeFilePathParts.push(relativeFilePathParts.pop().slice(0, -this.extension.length) + ".html");
      relativeFilePath = sysPath.join.apply(this, relativeFilePathParts.slice(1));
      newpath = sysPath.join(relAssetPath, relativeFilePath);
      return newpath;
    };

    StaticJadeCompiler.prototype.onCompile = function(changedFiles) {
      var _this = this;
      return changedFiles.every(function(file) {
        var f, filesToCompile, jadeFileName, newFilePath, _i, _len, _results;
        filesToCompile = (function() {
          var _i, _len, _ref, _results;
          _ref = file.sourceFiles;
          _results = [];
          for (_i = 0, _len = _ref.length; _i < _len; _i++) {
            f = _ref[_i];
            if (StaticJadeCompiler.prototype.isFileToCompile(f.path)) {
              _results.push(f.path);
            }
          }
          return _results;
        })();
        _results = [];
        for (_i = 0, _len = filesToCompile.length; _i < _len; _i++) {
          jadeFileName = filesToCompile[_i];
          newFilePath = StaticJadeCompiler.prototype.getHtmlFilePath(jadeFileName, _this.relAssetPath);
          try {
            _results.push(fromJade2Html(jadeFileName, _this.config, fileWriter(newFilePath)));
          } catch (err) {
            logError(err);
            _results.push(null);
          }
        }
        return _results;
      });
    };

    return StaticJadeCompiler;

  })();

}).call(this);
